libcontainer/rootfs_linux.go

```diff
@@ -733,14 +733,20 @@ func checkProcMount(rootfs, dest string, m *mountEntry) error {
 		if m == nil {
 			return nil
 		}
-		// only allow a mount on-top of proc if it's source is "proc"
-		st, err := m.srcStatfs()
-		if err != nil {
-			return err
-		}
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// only allow a mount on-top of proc if its source is a procfs mount ...
+		if m.Device == "proc" {
 			return nil
 		}
+		// ... or if the mount source is a bind-mount of a procfs mount.
+		if m.IsBind() {
+			st, err := m.srcStatfs()
+			if err != nil {
+				return err
+			}
+			if st.Type == unix.PROC_SUPER_MAGIC {
+				return nil
+			}
+		}
 		return fmt.Errorf("%q cannot be mounted because it is not of type proc", dest)
```

Comment:
It seems this line might not be accurate with this commit now, as the source could be of type proc and just not contain the "bind" option.

I'm unsure if we are improving things with this commit, it's not obvious at least.

Reply:
"is it not of type proc", is correct in both cases....

Reply:
To be clear, the issue I am fixing here is that `checkProcMount` doesn't currently return an error if you add (for example) an `overlay` mount that has a "source" of `/proc` but actually has a different `lowerdir`.

Hopefully that makes it more obvious that this clearly needs to be fixed.